packages/server/migrations/20240430210546_grants_viewed_unique_add_user_id.js

```diff
@@ -0,0 +1,48 @@
+// We want to be able to track whether a grant is viewed on a per-user basis (plus, ensure a grant
+// doesn't lose its viewed status when a user is deleted, and avoid 500s when a second user views),
+// so we're adding user_id to the compound PK for the table.
+
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+exports.up = async function (knex) {
+    // Knex builder doesn't support `CONCURRENTLY` or `USING INDEX` as options, so we use raw SQL
+    await knex.raw(`
+        CREATE UNIQUE INDEX CONCURRENTLY idx_grants_viewed_grant_id_agency_id_user_id_unique 
+            ON grants_viewed (grant_id, agency_id, user_id)
+    `);
+    await knex.raw(`
+        ALTER TABLE grants_viewed 
+            DROP CONSTRAINT grants_viewed_pkey,
+            ADD CONSTRAINT grants_viewed_pkey PRIMARY KEY USING INDEX idx_grants_viewed_grant_id_agency_id_user_id_unique;  
+    `);
+};
+
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+exports.down = async function (knex) {
+    // Rebuild the original index then flip the PK back to it
+    await knex.raw(`
+        CREATE UNIQUE INDEX CONCURRENTLY idx_grants_viewed_grant_id_agency_id_unique 
+            ON grants_viewed (grant_id, agency_id)
+    `);
+    await knex.raw(`
+        ALTER TABLE grants_viewed 
+            DROP CONSTRAINT grants_viewed_pkey,
+            ADD CONSTRAINT grants_viewed_pkey PRIMARY KEY USING INDEX idx_grants_viewed_grant_id_agency_id_unique;  
+    `);
+    // In the forward migration, user_id gets automatically marked non-nullable when it's included
+    // in the PK so we want to reverse that here
+    await knex.schema.table('grants_viewed', (table) => {
+        table.setNullable('user_id');
+    });
```

Comment:
I'm not sure it matters much, but since this seems likely to fail once a grant has been viewed twice for the same agency, I wonder if it's better to skip the downward revision here and instead implement the upward revision so that it's a no-op if there's nothing to do.

Reply:
Having the down migration was super-helpful in development, since I needed to revert this migration before switching back to another branch (or blow away the DB, but then I would need to re-load the staging dump). And it could conceivably be useful if there was a need to revert the deploy ~immediately, before any conflicting grant views were recorded. But yeah, you're totally right this would fail once you have multiple views recorded on a grant. 

What are the circumstances and likelihood the down migration would be used in production? (Should we raise an error to ensure we can't migrate down; should it silently noop so we can migrate down; should we drop duplicate rows and then restore the table to its original structure?)

It sounds like migrations are managed by hand (as opposed to being automatically run, and automatically rolled back for e.g. failed deploys), correct? If so, just a noop down migration would give us the ability to down-migrate through this to other down migrations, if the need arose. And if any issues arose with this PR/migration specifically, we could write up a quick fix PR expressly to address the issue. Is that what you were thinking? 

If so, I can update the down migration to a noop and the up migration to first check whether its changes have been applied so it can be up-migrated repeatedly without failure. But I wanted to check first, since it'll take a tiny bit of digging/testing to make sure the "check before applying the up migration" part works as expected.

Reply:
I'm not particularly worried about the down migration being run in Production environments (and yes, those are executed manually); this was more a concern focused on dev workflows that involve migrating the DB up, adding data, and then downgrading the DB to a state that's incompatible with the added data. It's enough of an edge-case concern that I'm not sure it's worth addressing proactively, especially given your report that having a working down migration path was quite useful for your own workflows.

Also, my own dev workflow involves regularly nuking and reloading the DB, so maybe that's skewing my judgement here :)

Reply:
Yeah ok, I agree there's a more technically correct approach than what's currently in this PR, but I also agree it's probably not worth any more time to implement. I'm going to move forward with landing this PR, thanks for talking that through!